vimiv/imutils/exif.py

```diff
@@ -11,138 +11,267 @@
 """
 
 import contextlib
+from typing import Dict
 
 from vimiv.utils import log, lazy
 from vimiv import api
 
+pyexiv2 = lazy.import_module("pyexiv2", optional=True)
 piexif = lazy.import_module("piexif", optional=True)
-
 _logger = log.module_logger(__name__)
 
 
-def check_piexif(return_value=None):
-    """Decorator for functions that require the optional piexif module.
+def check_exif_dependancy(return_value=None, check_piexif=True):
+    """Decorator for ExifHandler which require the optional py3exiv2 module.
 
-    If piexif is not available, return_value is returned and a debug log message is
-    logged. It it is available, the function is called as usual.
+    If py3exiv2 is available the class is left as it is. If py3exiv2 is not available
+    but the depreciated piexif module is, a depreciation warning is given to the user
+    and a ExifHandler class supporting piexif is returned. If none of the two modules
+    is available, a dummy class is returned and a debug log is logged.
 
     Args:
-        return_value: Value to return if piexif is not available.
+        return_value: Value to return if neither py3exiv2 nor piexif is available.
+        check_piexif: Check piexif dependency too in case py3exiv2 is not available.
     """
 
-    def decorator(func):
-        def stub(*_args, **_kwargs):
-            """Dummy function to call if piexif is not available."""
-            _logger.debug(
-                "Cannot call '%s', piexif is required for exif support", func.__name__
-            )
+    def decorator(handler):
+        if pyexiv2:
+            return handler
+
+        if piexif and check_piexif:
+
+            class ExifHandlerPiexif(handler):
```

Comment:
I have a few questions here:
- Why does this inherit from the other handler? As far as I can tell, there is no shared code.
- Could we just define this class (as well as `NoExifHandler`) outside of the decorator, i.e. below the regular handler,  then to avoid this huge nesting?

Reply:
- There is theoretically not need for inheritance, since everything gets overwritten. The idea of this implementation was to get some kind of a *cheap* interface/abstract class (indicate that the both classes have the same functionality). But it is probably better to remove the inheritance completely or do it properly using abstract classes.

Reply:
Improved in 4fd97f9f4bdb73e583092ab49486144d4a08fb36

Reply:
Much nicer, thanks! If defining another class ever becomes a relevant performance issue, we have done a lot of other things right :laughing: